### ISSUE TITLE
Incompatible generated LLVM type for character type 

### ISSUE THREAD

**Issue by serge-sans-paille** [lfortran/lfortran#4235]
The following

```
program char_cmp
    character b
    logical lsame
    external lsame

    if (lsame(b, 'b')) then
        print *, b
    end if

end program char_cmp
```
compiled with 

```
lfortran   --fixed-form-infer --cpp-infer --implicit-interface --generate-object-code -fPIC -c /tmp/a.f90 -o/dev/null 2>&1 | less
```

produces the error below:

```
; ModuleID = 'LFortran'
source_filename = "LFortran"

@0 = private unnamed_addr constant [2 x i8] c"b\00", align 1
@1 = private unnamed_addr constant [2 x i8] c" \00", align 1
@2 = private unnamed_addr constant [2 x i8] c"\0A\00", align 1
@3 = private unnamed_addr constant [5 x i8] c"%s%s\00", align 1

declare i1 @lsame(i8*, i8*)

define i32 @main(i32 %0, i8** %1) {
.entry:
  call void @_lpython_call_initial_functions(i32 %0, i8** %1)
  %b = alloca i8*, align 8
  %2 = call i8* @_lfortran_malloc(i32 2)
  call void @_lfortran_string_init(i32 2, i8* %2)
  store i8* %2, i8** %b, align 8
  %3 = load i8*, i8** %b, align 8
  %4 = call i1 @lsame(i8** %b, i8* getelementptr inbounds ([2 x i8], [2 x i8]* @0, i32 0, i32 0))
  br i1 %4, label %then, label %else

then:                                             ; preds = %.entry
  %5 = load i8*, i8** %b, align 8
  call void (i8*, ...) @_lfortran_printf(i8* getelementptr inbounds ([5 x i8], [5 x i8]* @3, i32 0, i32 0), i8* %5, i8* getelementptr inbounds ([2 x i8], [2 x i8]* @2, i32 0, i32 0))
  br label %ifcont

else:                                             ; preds = %.entry
  br label %ifcont

ifcont:                                           ; preds = %else, %then
  ret i32 0
}

declare void @_lpython_call_initial_functions(i32, i8**)

declare i8* @_lfortran_malloc(i32)

declare void @_lfortran_string_init(i32, i8*)

declare void @_lfortran_printf(i8*, ...)
asr_to_llvm: module failed verification. Error:
Call parameter type does not match function signature!
  %b = alloca i8*, align 8
 i8*  %4 = call i1 @lsame(i8** %b, i8* getelementptr inbounds ([2 x i8], [2 x i8]* @0, i32 0, i32 0))

code generation error: asr_to_llvm: module failed verification. Error:
Call parameter type does not match function signature!
  %b = alloca i8*, align 8
 i8*  %4 = call i1 @lsame(i8** %b, i8* getelementptr inbounds ([2 x i8], [2 x i8]* @0, i32 0, i32 0))

```

`%b` as type `i8**` but `i8*` is expected.

**Comment by serge-sans-paille** [comment]
I *think* we should have something along those lines instead:

```
  %b = alloca i8*, align 8
  %2 = call i8* @_lfortran_malloc(i32 2)
  call void @_lfortran_string_init(i32 2, i8* %2)
  store i8* %2, i8** %b, align 8
  %3 = load i8*, i8** %b, align 8
  %4 = call i1 @lsame(i8* %3, i8* getelementptr inbounds ([2 x i8], [2 x i8]* @0, i32 0, i32 0))
```

@certik can you confirm? If so I can work on a patch.

**Comment by certik** [comment]
A simpler way to reproduce:
```
 lfortran --implicit-interface a.f90
```
First thing to do is to check ASR via `--show-asr`, but it looks good to me. Then you can print Fortran source code from ASR using `lfortran --implicit-interface a.f90 --show-fortran`:
```fortran
program char_cmp
implicit none
character(len=1, kind=1) :: b
if (lsame(b, "b")) then
    print *, b
end if

contains

logical(4) function lsame(lsame_arg_0, lsame_arg_1) bind(c) result(lsame_return_var_name)
    character(len=1, kind=1) :: lsame_arg_0
    character(len=1, kind=1) :: lsame_arg_1
end function lsame

end program char_cmp
```
There is a slight mistake in our printer, it should actually be an interface. However, even with this code we can trigger the bug:
```console
$ lfortran b.f90
...
code generation error: asr_to_llvm: module failed verification. Error:
Call parameter type does not match function signature!
  %b = alloca i8*, align 8
 i8*  %4 = call i1 @lsame(i8** %b, i8* getelementptr inbounds ([2 x i8], [2 x i8]* @0, i32 0, i32 0))
```
The next step is to create a minimal reproducible example (MRE). We remove code while still triggering this bug. I got:
```fortran
program char_passing
character(1) :: b
b = "X"
call f(b)

contains

    subroutine f(s) bind(c)
    character(1) :: s
    if (s /= "X") error stop
    end subroutine

end program
```
I made it a little test as well that you can use for integration_tests. The bug is in `bind(c)`. Without it it works.

The function signature is correct:
```llvm
define void @f(i8* %s) {
```
But calling it is incorrect:
```llvm
  call void @f(i8** %b)
```
Instead, we need to "load" the pointer `b` before calling it. Exactly as you wrote above!

> @certik can you confirm?

So yes, I agree! Thanks. Let me know if you have any other questions.

**Comment by serge-sans-paille** [comment]
Fixed by #4243
